core/include/xfa.h

```diff
@@ -68,8 +68,8 @@ _Pragma("GCC diagnostic ignored \"-Warray-bounds\"")
 #define XFA_INIT_CONST(type, name) \
     _Pragma("GCC diagnostic push") \
     _Pragma("GCC diagnostic ignored \"-Wpedantic\"") \
-    _XFA_CONST(name, 0_) const volatile type name [0] = {}; \
-    _XFA_CONST(name, 9_) const volatile type name ## _end [0] = {}; \
+    _XFA_CONST(name, 0_) const type name [0] = {}; \
```

Comment:
the volatile was needed, what's the reason for removing?

Reply:
With `volatile` my `gcc` complains about a type mismatch, when using `XFA_INIT_CONST` and `XFA_USE_CONST`.

<details>
<summary>gcc</summary>
<p>

```
In file included from /home/fabian/forks/RIOT/sys/include/auto_init_macros.h:24,
                 from /home/fabian/forks/RIOT/sys/auto_init/auto_init.c:27:
/home/fabian/forks/RIOT/sys/include/auto_init_macros.h:415:33: error: conflicting types for 'auto_init_xfa'; have 'void (* const[])(void)'
  415 |     XFA_USE_CONST(auto_init_fn, auto_init_xfa);                                         \
      |                                 ^~~~~~~~~~~~~
/home/fabian/forks/RIOT/core/include/xfa.h:110:23: note: in definition of macro 'XFA_USE_CONST'
  110 |     extern const type name []; \
      |                       ^~~~
/home/fabian/forks/RIOT/sys/auto_init/auto_init.c:33:5: note: in expansion of macro 'AUTO_INIT_MODULE'
   33 |     AUTO_INIT_MODULE(ztimer, AUTO_INIT_PRIO_ZTIMER, ztimer_init);
      |     ^~~~~~~~~~~~~~~~
/home/fabian/forks/RIOT/sys/auto_init/auto_init.c:30:30: note: previous definition of 'auto_init_xfa' with type 'void (* const volatile[0])(void)' {aka 'void (* const volatile[])(void)'}
   30 | XFA_INIT_CONST(auto_init_fn, auto_init_xfa);
      |                              ^~~~~~~~~~~~~
/home/fabian/forks/RIOT/core/include/xfa.h:71:46: note: in definition of macro 'XFA_INIT_CONST'
   71 |     _XFA_CONST(name, 0_) const volatile type name [0] = {}; \
      |                                              ^~~~
/home/fabian/forks/RIOT/sys/include/auto_init_macros.h:415:33: error: conflicting types for 'auto_init_xfa_end'; have 'void (* const[])(void)'
  415 |     XFA_USE_CONST(auto_init_fn, auto_init_xfa);                                         \
      |                                 ^~~~~~~~~~~~~
/home/fabian/forks/RIOT/core/include/xfa.h:111:23: note: in definition of macro 'XFA_USE_CONST'
  111 |     extern const type name ## _end []
      |                       ^~~~
/home/fabian/forks/RIOT/sys/auto_init/auto_init.c:33:5: note: in expansion of macro 'AUTO_INIT_MODULE'
   33 |     AUTO_INIT_MODULE(ztimer, AUTO_INIT_PRIO_ZTIMER, ztimer_init);
      |     ^~~~~~~~~~~~~~~~
/home/fabian/forks/RIOT/sys/auto_init/auto_init.c:30:30: note: previous definition of 'auto_init_xfa_end' with type 'void (* const volatile[0])(void)' {aka 'void (* const volatile[])(void)'}
   30 | XFA_INIT_CONST(auto_init_fn, auto_init_xfa);
      |                              ^~~~~~~~~~~~~
/home/fabian/forks/RIOT/core/include/xfa.h:72:46: note: in definition of macro 'XFA_INIT_CONST'
   72 |     _XFA_CONST(name, 9_) const volatile type name ## _end [0] = {}; \
      |                                              ^~~~
/home/fabian/forks/RIOT/sys/include/auto_init_macros.h:415:33: error: conflicting types for 'auto_init_xfa'; have 'void (* const[])(void)'
  415 |     XFA_USE_CONST(auto_init_fn, auto_init_xfa);                                         \
      |                                 ^~~~~~~~~~~~~
/home/fabian/forks/RIOT/core/include/xfa.h:110:23: note: in definition of macro 'XFA_USE_CONST'
  110 |     extern const type name []; \
      |                       ^~~~
/home/fabian/forks/RIOT/sys/auto_init/auto_init.c:41:5: note: in expansion of macro 'AUTO_INIT_MODULE'
   41 |     AUTO_INIT_MODULE(random, AUTO_INIT_PRIO_RANDOM, auto_init_random);
      |     ^~~~~~~~~~~~~~~~
/home/fabian/forks/RIOT/sys/auto_init/auto_init.c:30:30: note: previous definition of 'auto_init_xfa' with type 'void (* const volatile[0])(void)' {aka 'void (* const volatile[])(void)'}
   30 | XFA_INIT_CONST(auto_init_fn, auto_init_xfa);
      |                              ^~~~~~~~~~~~~
/home/fabian/forks/RIOT/core/include/xfa.h:71:46: note: in definition of macro 'XFA_INIT_CONST'
   71 |     _XFA_CONST(name, 0_) const volatile type name [0] = {}; \
      |                                              ^~~~
/home/fabian/forks/RIOT/sys/include/auto_init_macros.h:415:33: error: conflicting types for 'auto_init_xfa_end'; have 'void (* const[])(void)'
  415 |     XFA_USE_CONST(auto_init_fn, auto_init_xfa);                                         \
      |                                 ^~~~~~~~~~~~~
/home/fabian/forks/RIOT/core/include/xfa.h:111:23: note: in definition of macro 'XFA_USE_CONST'
  111 |     extern const type name ## _end []
      |                       ^~~~
/home/fabian/forks/RIOT/sys/auto_init/auto_init.c:41:5: note: in expansion of macro 'AUTO_INIT_MODULE'
   41 |     AUTO_INIT_MODULE(random, AUTO_INIT_PRIO_RANDOM, auto_init_random);
      |     ^~~~~~~~~~~~~~~~
/home/fabian/forks/RIOT/sys/auto_init/auto_init.c:30:30: note: previous definition of 'auto_init_xfa_end' with type 'void (* const volatile[0])(void)' {aka 'void (* const volatile[])(void)'}
   30 | XFA_INIT_CONST(auto_init_fn, auto_init_xfa);
      |                              ^~~~~~~~~~~~~
/home/fabian/forks/RIOT/core/include/xfa.h:72:46: note: in definition of macro 'XFA_INIT_CONST'
   72 |     _XFA_CONST(name, 9_) const volatile type name ## _end [0] = {}; \
      |                                              ^~~~
/home/fabian/forks/RIOT/sys/include/auto_init_macros.h:415:33: error: conflicting types for 'auto_init_xfa'; have 'void (* const[])(void)'
  415 |     XFA_USE_CONST(auto_init_fn, auto_init_xfa);                                         \
      |                                 ^~~~~~~~~~~~~
/home/fabian/forks/RIOT/core/include/xfa.h:110:23: note: in definition of macro 'XFA_USE_CONST'
  110 |     extern const type name []; \
      |                       ^~~~
/home/fabian/forks/RIOT/sys/auto_init/auto_init.c:207:5: note: in expansion of macro 'AUTO_INIT_MODULE'
  207 |     AUTO_INIT_MODULE(test_utils_interactive_sync, AUTO_INIT_PRIO_TEST_UTILS_INTERACTIVE_SYNC, test_utils_interactive_sync);
      |     ^~~~~~~~~~~~~~~~
/home/fabian/forks/RIOT/sys/auto_init/auto_init.c:30:30: note: previous definition of 'auto_init_xfa' with type 'void (* const volatile[0])(void)' {aka 'void (* const volatile[])(void)'}
   30 | XFA_INIT_CONST(auto_init_fn, auto_init_xfa);
      |                              ^~~~~~~~~~~~~
/home/fabian/forks/RIOT/core/include/xfa.h:71:46: note: in definition of macro 'XFA_INIT_CONST'
   71 |     _XFA_CONST(name, 0_) const volatile type name [0] = {}; \
      |                                              ^~~~
/home/fabian/forks/RIOT/sys/include/auto_init_macros.h:415:33: error: conflicting types for 'auto_init_xfa_end'; have 'void (* const[])(void)'
  415 |     XFA_USE_CONST(auto_init_fn, auto_init_xfa);                                         \
      |                                 ^~~~~~~~~~~~~
/home/fabian/forks/RIOT/core/include/xfa.h:111:23: note: in definition of macro 'XFA_USE_CONST'
  111 |     extern const type name ## _end []
      |                       ^~~~
/home/fabian/forks/RIOT/sys/auto_init/auto_init.c:207:5: note: in expansion of macro 'AUTO_INIT_MODULE'
  207 |     AUTO_INIT_MODULE(test_utils_interactive_sync, AUTO_INIT_PRIO_TEST_UTILS_INTERACTIVE_SYNC, test_utils_interactive_sync);
      |     ^~~~~~~~~~~~~~~~
/home/fabian/forks/RIOT/sys/auto_init/auto_init.c:30:30: note: previous definition of 'auto_init_xfa_end' with type 'void (* const volatile[0])(void)' {aka 'void (* const volatile[])(void)'}
   30 | XFA_INIT_CONST(auto_init_fn, auto_init_xfa);
      |                              ^~~~~~~~~~~~~
/home/fabian/forks/RIOT/core/include/xfa.h:72:46: note: in definition of macro 'XFA_INIT_CONST'
   72 |     _XFA_CONST(name, 9_) const volatile type name ## _end [0] = {}; \
      |                                              ^~~~
make[3]: *** [/home/fabian/forks/RIOT/Makefile.base:142: /home/fabian/forks/RIOT/tests/auto_init_custom/bin/nucleo-f767zi/auto_init/auto_init.o] Fehler 1
make[2]: *** [/home/fabian/forks/RIOT/Makefile.base:31: ALL--/home/fabian/forks/RIOT/sys/auto_init] Fehler 2
make[1]: *** [/home/fabian/forks/RIOT/Makefile.base:31: ALL--/home/fabian/forks/RIOT/sys] Fehler 2
make: *** [/home/fabian/forks/RIOT/tests/auto_init_custom/../../Makefile.include:686: application_tests_auto_init_custom.module] Fehler 2
```

</p>
</details>  

```
$ /usr/bin/arm-none-eabi-gcc --version
arm-none-eabi-gcc (Arch Repository) 11.2.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

```

However, if both use `volatile` it´s also fine, if that would be better.

Reply:
Well, ideally the volatile is not needed, but the msp430 gcc optimized too much without volatile. We'll need to double-check.
Do you need to use both XFA_INIT_CONST and XFA_USE_CONST in the same compilation unit?

Reply:
I guess a way to not use XFA_INIT_CONST and XFA_USE_CONST in the same compilation unit would be if
`XFA_INIT_CONST(auto_init_fn, auto_init_xfa)` gets its very own source file `auto_init_xfa.c`.  That would be ok. If not, I can also think of something like this:

```
#if defined(MSP430)
XFA_INIT_CONST(volatile auto_init_fn, auto_init_xfa);
#else
XFA_INIT_CONST(auto_init_fn, auto_init_xfa);
#endif

Reply:
no, `XFA_INIT_CONST(auto_init_fn, auto_init_xfa)` can stay in auto_init.c.

I think the `AUTO_INIT_MODULE` macros don't need the declarator, see https://github.com/RIOT-OS/RIOT/pull/17109/files#r741798147.

If that declarator is not needed, that would solve this I think?

Reply:
yes